### DISCUSSION TITLE
Integrate with Weblate

### DISCUSSION THREAD

**Discussion by JakobDev** [ximion/appstream#473] [Q&A]
I'm currently setting up Weblate for my Projects. I also want to include the AppStream and my .desktop (It looks like I can do this using xgettext) file. As Weblate does not support both directly, I need to create a other format. The [Documentation ](https://www.freedesktop.org/software/appstream/docs/sect-Quickstart-Translation.html)states 3 Ways of doing this: intltool, a autotools macro and itstool. All Ways uses gettext, which I have never used before.

I failed to got intltool working, and even if I could, I need to change the structure of the file (use underscores) which can lead to problems.

As I don't use autotools, I can't use the macro.

The only remaining Way is itstool and I got it working, but it has big problem: Every time I run `itstool -i as-metainfo.its -o translations/test.po test.metainfo.xml` it deletes all translations. This means, when I add a new Screenshot with a Caption and want to add the new Caption to Weblate, I'm losing all translations, so this is not useable for me.

Is there a preferred Way to integrate AppStream with Weblate?

**Comment by tintou** [comment]
It will depend on the build system you're using, when using meson for instance you can use the i18n plugin available to do something like this https://github.com/elementary/calculator/blob/master/po/extra/meson.build (it also requires a [LINGUAS](https://github.com/elementary/calculator/blob/master/po/extra/LINGUAS) and [POTFILES](https://github.com/elementary/calculator/blob/master/po/extra/POTFILES) to work)

**Reply by JakobDev** [reply]
I use mostly Python, so I don't really use a build system at all beside setuptools, which is really not capable of this things. I'm currently trying to write 2 scripts: One which creates and updates the files for Weblate and one which merges them.

**Reply by tintou** [reply]
Yes then that is the way to go, you can conveniently look at how meson itself is doing it https://github.com/mesonbuild/meson/blob/master/mesonbuild/modules/i18n.py as it is done in python too :)
The `gettext` method is the one updating the template file and `merge_file` is the one applying the translation

**Comment by ximion** [comment]
The MetaInfo Creator will generate Meson snippets for you to translate the files: https://www.freedesktop.org/software/appstream/metainfocreator/#/
Ans AppStream itself also translates its files. But AppStream makes sure the release notes are not translated, which is something you may actually want in your project.
For something like AppStream, when using xgettext for your project already, you just need to add the metainfo file to your `POTFILES.in` file: https://github.com/ximion/appstream/blob/master/po/POTFILES.in#L4
And then you can get a translated file using itstool: https://github.com/ximion/appstream/blob/master/data/meson.build#L20
You don't need the `its_files` parameter for your project, Itstool and Gettext will automatically do the right things as AppStream installs ITS files for them (only AppStream itself needs to work without the ITS files being installed).

And that should pretty much be it :-)

**Reply by buhtz** [reply]
Never used Meason but seems to be a build system like make. But in Python universe you don't need an extern build system. Python packages can build (nearly) themself. I don't want to introduce a new build system just for translation. The syntax is not self describing also.

But it seems you do use "itstool" under the hood?

How do you make sure that your POTFILES.in is up to date? Why not generating the file list on the fly?

**Comment by JakobDev** [comment]
As already set, I don't use gettext or meson. I have now accomplished to create a Example Project, but I'm not sure, if there are better Ways to do this, as I don't have any experience in gettext.
[translation.zip](https://github.com/ximion/appstream/files/10974375/translation.zip)

**Reply by ximion** [reply]
There's no real way to do this without Gettext - if you at least have pot files, you could use gettext only for the metainfo file, create a pot file and then merge that into your original file.
If you don't even use pot files, I don't know how you'd do this, you would likely need to write some custom tooling in that case.

**Reply by JakobDev** [reply]
The attached example actually uses gettext with custom scripts

**Comment by buhtz** [comment]
I found a solution but did not implement it yet. Have look at this two entities in my repo when looking for a solution:
 - https://github.com/bit-team/backintime/blob/dev/update_language_files.py
 - https://github.com/bit-team/backintime/blob/dev/common/doc-dev/2_localization.md

In short. You can use the regular GNU gettext tools. Nothing else needed.
Extract translatable strings (via xgettext) into a pot-file. Concat that pot file with other pot-files you might have (e.g. from your python sources).
From here it is the usual: Generate/Update po-files for each language. Upload them (somehow) to Weblate.
When translation is finished it seems you have to use "itstool" after generating mo files out of the po. It seems that "msgfmt" do not support an xml/appstream operation mode yet.

**Reply by buhtz** [reply]
Can mark my own answer as answer. 🤣

**Reply by ximion** [reply]
> In short. You can use the regular GNU gettext tools. Nothing else needed.

Careful if you have any `code`/`em` etc. markup in your descriptions - xgettext does not handle these well and may just erase them (itstool is fine). Apart from that, I think this should work :-)
